### ISSUE TITLE
feat: add sharding rules API and CLI

### ISSUE THREAD

**Issue by pauldix** [influxdata/influxdb_iox#919]
EDIT: we opted to define the routing rules in terms of the sharding semantics the user desire, where the request routing is just a means to the end of achieving such sharding semantics.

---

This is part of delivering #916. The routing rules defined in #918 need to be configurable by the gRPC management API. We'll need to be able to define a route, which itself will need the list of specific matching rules and the default match (fallback).

* Create/set a route (with given name)
  * Should validate that the server knows about the passed in IOx writer IDs
* Delete a route (with given name)

**Possible CLI**
Should this be a subcommand of database because it's always within the scope of a DB?

```
# create a route named asdf for database foo. If you want to re-order the rules, re-create the route
./influxdb_iox database foo route create asdf -d '{rules: [{table: "cpu" node: 23}], default: {hash_table: true, hash_columns: ["host"], nodes: [23, 42, 4, 77]}}'

# list the routes (returns an array of json structs like the one above in -d
./influxdb_iox database foo route list

# delete a specific route
./influxdb_iox database foo route delete asdf
```

I'm guessing the passing of the JSON data might spark some debate. This could be broken out into subcommands, but it would add quite a bit to the surface area. Also, the ordering of the rules is important so users would end up having to call adding rules in the correct order.

**Comment by alamb** [comment]
@pauldix  I am thinking that we shouldn't add the ability to configure every conceivable hierarchical structure via a flat CLI yet. I suggest the CLI should focus on "basic administrative tasks that a human is likely to do" and anything more sophisticated we punt to something like JSON

In other words, I think the JSON is fine for now (and we can make a full featured flattened CLI at some point in the future once the data structured is nailed down and that we have users who want it)

**Comment by mkmik** [comment]
tentative CLI

```shell
# configure a shard for database foo
./influxdb_iox database shard set foo -d '{rules: [{table: "cpu" node: 23}], default: {hash_table: true, hash_columns: ["host"], nodes: [23, 42, 4, 77]}}'

# list the shards (returns an array of json structs like the one above in -d)
./influxdb_iox database shard list

# delete sharding config for database foo
./influxdb_iox database shard remove foo
```

**Comment by pauldix** [comment]
This should be ready to get picked up. Although the Shard configuration will likely have to be updated in the `DatabaseRules` and proto. The `ShardConfig` will need a mapping of `ShardId` to `NodeGroup` and the specific target will need to map to a `ShardId` rather than a node group. And the `HashRing` will need a ring of `ShardId` rather than node group.

And `ShardConfig` will have to implement the `Sharder` interface.

**Comment by mkmik** [comment]
I'm in it, blocked on #1066

**Comment by mkmik** [comment]
@pauldix could you please create a new ticket to track the shardconfig refactoring? This is about API and CLI
